src/libutil/util.cc

```diff
@@ -786,13 +786,64 @@ pid_t Pid::release()
     return p;
 }
 
-
 void killUser(uid_t uid)
 {
     debug(format("killing all processes running under uid '%1%'") % uid);
 
     assert(uid != 0); /* just to be safe... */
 
+#ifdef __APPLE__
+    /* We originally had very similar logic here as on the Linux side, but
+       then got bitten hard by a system crashing bug as described in here:
+       https://github.com/NixOS/nix/issues/1583. So now we adopt this other
+       approach, but we should go back to the kill approach once Apple fixes
+       their kernel. */
+
+    // Borrowed from https://stackoverflow.com/a/7733928
+    int err = 0;
+    size_t length = 0, newlength = 0;
+
+    // XXX: Should this be KERN_PROC_RUID? EUID vs. real UID
+    static const int name[4] = { CTL_KERN, KERN_PROC, KERN_PROC_UID, static_cast<int>(uid) };
+
+    int tries = 100;
+    bool done = false;
+
+    while (!done) {
+        tries--;
+
+        // First call with NULL to get number of processes. Yeah yeah TOCTOU is fun, but we check
+        // that it hasn't changed later and keep going until it's 0
```

Comment:
We don't check that it hasn't changed later, right?

Reply:
Yeah I don't. Should I? My first thought was that I'd need to, but then I thought some more and convinced myself I wouldn't.

Reply:
Nah, shouldn't be necessary, just need to update the comment